### PR TITLE
Fix performance of cosmetic filters

### PR DESCRIPTION
**Pull request by atuchin-m**
For https://github.com/brave/brave-browser/issues/22030
1.37.5 => good
image (13)

1.37.6 => bad
image (12)

Try to revert https://github.com/brave/brave-core/pull/12038 that makes cosmetic filters slower.
## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

See https://github.com/brave/brave-browser/issues/22030#issuecomment-1090632101

### CODE REVIEW

**Comment by SergeyZhukovsky** [comment]
@atuchin-m nice find. Need perhaps to polish and profile more what @antonok-edm did before we can re-merge it back.

**Comment by SergeyZhukovsky** [comment]
we have also reports on Android that 1.37.x is slower comparing to 1.36.x

**Review by SergeyZhukovsky — APPROVED** [review]
need to fix a few tests before merge, but otherwise looks good

File: browser/brave_shields/ad_block_service_browsertest.cc [hunk 1]
```diff
@@ -2158,7 +2153,6 @@ IN_PROC_BROWSER_TEST_F(AdBlockServiceTest, CosmeticFilteringIframeScriptlet) {
   ASSERT_EQ(true, EvalJs(contents, "show_ad"));
 }
 
-<<<<<<< HEAD
 // Test cosmetic filtering on an element that already has an `!important`
 // marker on its `display` style.
 IN_PROC_BROWSER_TEST_F(AdBlockServiceTest,
```
Comment (antonok-edm) [inline browser/brave_shields/ad_block_service_browsertest.cc:2158]: `AdBlockServiceTest.CosmeticFilteringOverridesImportant` will have to be disabled as well, that was the primary issue fixed by the original PR

Comment (atuchin-m) [reply browser/brave_shields/ad_block_service_browsertest.cc:2158]: @antonok-edm also CosmeticFilteringDynamicCustom doesn't work. Is it expected?

Comment (antonok-edm) [reply browser/brave_shields/ad_block_service_browsertest.cc:2158]: Yeah, unfortunately it looks like I inserted two different fixes into the same PR. If you un-revert the changes in `CosmeticFilteringDynamicCustom` and in `test/data/cosmetic_filtering.html`, I think that one should work fine.

Comment (atuchin-m) [reply browser/brave_shields/ad_block_service_browsertest.cc:2158]: In fact it's not enough..
```
[55080:56304:0411/175238.315:INFO:CONSOLE(5)] "still waiting for css selector", source: __const_std::string&_script__ (5)
content/public/test/browser_test_utils.cc(2659): error: Failed
RunLoop::Run() timed out. Timeout set at ProxyRunTestOnMainThreadLoop@content/public/test/browser_test_base.cc:809.
```

Comment (antonok-edm) [reply browser/brave_shields/ad_block_service_browsertest.cc:2161]: I just pushed a couple of fixes for that as well.

**Review by antonok-edm — APPROVED** [review]

**Comment by antonok-edm** [comment]
macOS CI failures are known/unrelated

**Comment by kjozwiak** [comment]
Reproduced the original issue on `Win 11 x64` using the STR/Cases outlined via https://github.com/brave/brave-browser/issues/22030#issuecomment-1090632101 with the following build:

```
Brave | 1.39.47 Chromium: 100.0.4896.79 (Official Build) nightly (64-bit)
-- | --
Revision | 8fb749dcab8700c24213791969e59deb72fee36f-refs/branch-heads/4896@{#1015}
OS | Windows 11 Version 21H2 (Build 22000.593)
```

* Shields disabled - `~250ms - 300ms`
* Shields enabled - `~850ms - 1s`

**`Shields Disabled`** | **`Shields Enabled`**
-------------------------|-------------------------
withoutShieldEnabled | withShieldsEnabled

**Comment by atuchin-m** [comment]
It should be >=1.39.49, 48 doesn't have the fix:
image

**Comment by kjozwiak** [comment]
> It should be >=1.39.49, 48 doesn't have the fix: image

yup, used `1.39.52 Chromium: 100.0.4896.88` 👍 It was technically supposed to be in `1.39.48` but looks like that used to basically bump the tags. Finished verifying and updated the above notes.

**Pull request merged**